client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java

```diff
@@ -568,7 +570,14 @@ static String buildEndpoint(String... parts) {
         StringJoiner joiner = new StringJoiner("/", "/", "");
         for (String part : parts) {
             if (Strings.hasLength(part)) {
-                joiner.add(part);
+                try {
+                    //encode each part (e.g. index, type and id) separately before merging them into the path
+                    URIBuilder uriBuilder = new URIBuilder().setPath(part);
+                    //make sure that "/" in each part are properly encoded too
+                    joiner.add(uriBuilder.build().getRawPath().replaceAll("/", "%2F"));
```

Comment:
Could we use a static encode() method for this? We'll have to reuse it in some URL parameters like routing or parent, right?

Reply:
it's complicated :) querystring params should already be encoded by the rest low-level client. but the low-level client expected paths to be externally encoded, as they are provided altogether and we can't encode each part safely anymore. In the high-level client though, it makes sense to encode each part separately as they are provided separately. I think what we do here is all we need.

Reply:
I will add tests for this.

Reply:
Of course, I just forgot that the low-level client already encodes parameters. Thanks.